### PR TITLE
https://cdap.atlassian.net/browse/CDAP-19368

### PR DESCRIPTION
**Pull request by udurbha**
After adding the column Names,if the path field value is entered adding pathfield value as column Name.
Added Tests for generating and extracting Pathfield column.

### CODE REVIEW

**Comment by udurbha** [comment]
Added pathfield addition code for along with columns when pathfield value is entered

**Comment by udurbha** [comment]
@albertshau

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java [hunk 1]
```diff
@@ -102,29 +102,33 @@ public static List<Schema.Field> detectDataTypeOfEachDatasetColumn(
    * @return Array of column names.
    */
   public static String[] setColumnNames(String rawLine, boolean skipHeader,
-                                        boolean enableQuotedValues, String delimiter) {
+                                        boolean enableQuotedValues, String delimiter, String pathField) {
```
Comment (albertshau) [inline format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java]: should not be passing the path field here, as it has nothing to do with data type detection. Adding it to the schema should be done at a different place

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java [hunk 2]
```diff
@@ -113,16 +113,21 @@ static Schema detectSchema(DelimitedConfig conf, String delimiter,
...
+
       try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputFile.open()))) {
         for (int rowIndex = 0; rowIndex < conf.getSampleSize() && (line = reader.readLine()) != null; rowIndex++) {
+
           rowValue = line.split(delimiter, -1);
           if (rowIndex == 0) {
+
             columnNames = DataTypeDetectorUtils.setColumnNames(line, conf.getSkipHeader(), conf.getEnableQuotedValues(),
-                                                               delimiter);
+                                                               delimiter, pathField);
```
Comment (albertshau) [inline format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java]: the path field is not related to detecting schema from data and should not be passed into the detector. Instead, it should be added to the list of fields that the detector returns.

File: format-delimited/src/test/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtilsTest.java [hunk 3]
```diff
@@ -38,47 +38,63 @@ public class DataTypeDetectorUtilsTest {
   @Test
   public void testExtractedColumnNames() {
     String headerLine = "column_A;column_B;column_C";
-    String[] actualColumnNames = DataTypeDetectorUtils.setColumnNames(headerLine, true, false, ";");
+    String[] actualColumnNames = DataTypeDetectorUtils.setColumnNames(headerLine, true, false, ";", null);
```
Comment (albertshau) [inline format-delimited/src/test/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtilsTest.java]: tests should be added to the format test classes (AvroInputFormatProviderTest, DelimitedInputFormatProviderTest, etc)

**Review by albertshau — COMMENTED** [review]
I'm pretty sure the avro and parquet formats still have the problem, please make sure you add tests for them and corresponding fixes if needed.

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java (hunk 1 above)
Comment (udurbha) [reply format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java]: Removed pathfield

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java [hunk 4]
```diff
@@ -111,25 +111,33 @@ static Schema detectSchema(DelimitedConfig conf, String delimiter,
...
       try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputFile.open()))) {
         for (int rowIndex = 0; rowIndex < conf.getSampleSize() && (line = reader.readLine()) != null; rowIndex++) {
+
           rowValue = line.split(delimiter, -1);
           if (rowIndex == 0) {
-            columnNames = DataTypeDetectorUtils.setColumnNames(line, conf.getSkipHeader(), conf.getEnableQuotedValues(),
-                                                               delimiter);
+
+            columnNamesList = DataTypeDetectorUtils.setColumnNames(line, conf.getSkipHeader(),
+                    conf.getEnableQuotedValues(), delimiter);
```
Comment (albertshau) [inline format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java:125]: indentation should be consistent with the rest of the project. In this case, it should align with the previous line. Please ensure your IDE is configured with the right settings.

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java [hunk 5]
```diff
@@ -46,15 +46,15 @@ public class DataTypeDetectorUtils {
    */
   public static void detectDataTypeOfRowValues(Map<String, Schema> override,
                                                DataTypeDetectorStatusKeeper dataTypeDetectorStatusKeeper,
-                                               String[] columnNames, String[] rowValues) {
+                                               List<String> columnNames, String[] rowValues) {
```
Comment (albertshau) [inline format-delimited/src/main/java/io/cdap/plugin/format/delimited/common/DataTypeDetectorUtils.java:49]: there should not be any changes in this class, it is unrelated to the path field.

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java [hunk 6]
```diff
@@ -111,25 +111,33 @@ static Schema detectSchema(DelimitedConfig conf, String delimiter,
...
+            if (conf.getPathField() != null) {
```
Comment (albertshau) [inline format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedInputFormatProvider.java:126]: should not add it here. Add it to the list of fields right before the schema is created

**Review by albertshau — COMMENTED** [review]
As mentioned previously, please add a proper test. It should be in DelimitedInputFormatProviderTest.

**Pull request closed without merging**